[Inductor] Exclude FusedMixOrderReductions from combo kernel grouping

karthickai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168050
* __->__ #169721
* #169509
* #168939

Fixes: #169811

This PR fixes group mismatched error `torch._inductor.exc.InductorError: NotImplementedError: unexpected group: (8192, 1024) != (65536, 128)`- occurs because `FusedMixOrderReductions` contains nodes with reversed groups that are incompatible with combo kernel's `generate_node_schedule()`


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

karthickai committed:
[WIP][Inductor] Exclude FusedMixOrderReductions from combo kernel grouping

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Exclude FusedMixOrderReductions from combo kernel grouping"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Exclude FusedMixOrderReductions from combo kernel grouping"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Exclude FusedMixOrderReductions from combo kernel grouping"


Fixes: #169811

This PR fixes group mismatched error `torch._inductor.exc.InductorError: NotImplementedError: unexpected group: (8192, 1024) != (65536, 128)`- occurs because `FusedMixOrderReductions` contains nodes with reversed groups that are incompatible with combo kernel's `generate_node_schedule()`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -2441,6 +2459,10 @@ def _default_group_nodes_for_combo_kernels(
                     continue
+
+                # exclude nodes that read from FusedMixOrderReductions output buffers'
+                if node.used_buffer_names() & excluded_buffer_names:
```
why can't we read from FusedMixOrderReductions

karthickai replied:
From the filtered nodes only FusedMixOrderReductions produces the intermediate buffer. FusedMixOrderReductions completes execution, its output buffers  are immediately marked for removal. When consumer nodes are combo-kerneled with other independent operations, the combo kernel is scheduled later, creating a timing gap where the buffers are already freed (V.graph.removed_buffers.add(name)) before the combo kernel tries to read them, resulting in AssertionError.

karthickai commented on torch/_inductor/scheduler.py:
```diff
@@ -2432,6 +2440,16 @@ def _default_group_nodes_for_combo_kernels(
         grouped_nodes = []
         max_num_nodes = 8
+
+        excluded_buffer_names: OrderedSet[str] = OrderedSet(
```
Even after excluding `FusedMixOrderReductions` itself from combo kernels, we will get the buffer removal error. This happens because `FusedMixOrderReductions` produces output buffers (e.g., buf0 for row_sum, buf2 for col_sum) which get marked as removed after the node completes execution. but, the consumers of these buffers nodes that read from `FusedMixOrderReductions` outputs may still be combo-kerneled together with other independent ops like `y.sum()` or` z.sum()`. When the combo kernel eventually runs and attempts to load these buffers, they have already been deleted, triggering the assertion failure. so, we added buffer filtering in `_default_group_nodes_for_combo_kernels()` (access to all topological level) that uses excluded_buffer_names to identify and exclude any nodes that read from FusedMixOrderReductions output buffers.

karthickai committed:
Update on "[Inductor] Exclude FusedMixOrderReductions from combo kernel grouping"


Fixes: #169811

This PR fixes group mismatched error `torch._inductor.exc.InductorError: NotImplementedError: unexpected group: (8192, 1024) != (65536, 128)`- occurs because `FusedMixOrderReductions` contains nodes with reversed groups that are incompatible with combo kernel's `generate_node_schedule()`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Exclude FusedMixOrderReductions from combo kernel grouping"


Fixes: #169811

This PR fixes group mismatched error `torch._inductor.exc.InductorError: NotImplementedError: unexpected group: (8192, 1024) != (65536, 128)`- occurs because `FusedMixOrderReductions` contains nodes with reversed groups that are incompatible with combo kernel's `generate_node_schedule()`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai commented:
@pytorchbot merge

karthickai commented:
@pytorchbot merge -i
